docs: add migration hub for Pyth Core to Pyth Pro upgrade (#3712)
pyth-network/pyth-crosschain: apps/developer-hub/content/docs/price-feeds/core/upgrade/index.mdx

---

> title: "Upgrading Pyth Core"

Reviewer: I like Preparing for the Pyth Core upgrade better.
Pyth Core is getting upgraded. They are not the ones upgrading Pyth Core.

Author: Already fixed. I will push that into next commit.

> description: "Everything you need to upgrade your Pyth Core integration before July 31, 2026."

Reviewer: ```suggestion
description: "Everything you need to upgrade your Pyth Core integration before August 31, 2026."
```
Double check all dates are coherent across documents.

Author: I intentionally kept July 31, so that we have some buffer.

  Choose: swap addresses now, or wait for the automatic upgrade (see the decision section below).
- **You only use a protocol that already integrates Pyth?**
  No action needed from your side.

## Your upgrade path

### Step 1: Get a Pyth API Key

Required for everyone who calls Hermes.
Sign up at Pyth Terminal: a free trial is included, paid plans cover ongoing use.

<Button
  href="https://pythdata.app/signup"
  rel="noopener noreferrer"
  size="md"
  target="_blank"
  variant="primary"
>
  Sign up at Pyth Terminal
</Button>

### Step 2: Upgrade now or wait for automatic?

After Step 1, choose how and when to move to the upgraded infrastructure. Your choice is saved in the URL so you can share or bookmark a specific path.

<BranchToggle />

|                  | Upgrade now (recommended)                  | Wait for automatic                                                   |
| ---------------- | ------------------------------------------ | -------------------------------------------------------------------- |
| Timing           | You choose                                 | July 31, 2026                                                        |

> | Downtime         | None                                       | Brief, during the switch                                             |

Reviewer: ```suggestion
| Downtime         | Up to you                                       | Brief, during the switch                                             |
```

> | Hermes endpoint  | Switch to the upgraded URL now             | Push `Bearer` auth on `hermes.pyth.network` at cutover               |

Reviewer: ```suggestion
| Hermes endpoint  | Switch to the upgraded URL now             | Start using `hermes.pyth.network` with an API key before the cutover               |
```

Reviewer: For a more smooth transition, they should start calling Hermes with the API key as soon as possible rather than scrambling to add the API key it at the exact cutover.

> | Contract address | You swap to the upgraded Pyth Core Contract       | DAO upgrades it for you                                              |

Reviewer: ```suggestion
| Contract address | You swap to the upgraded Pyth Core Contract       | DAO upgrades the current Pyth Core Contract for you                             |
```

<BranchSection path="now">

> **Upgrade now (recommended)**: Continue with Steps 2 and 3 below. You move to the [upgraded Hermes endpoint](https://pyth.dourolabs.app/hermes) and the [upgraded Pyth Core Contract](/price-feeds/core/upgrade/contracts) today, on your own schedule, with zero downtime.

Reviewer: It's only zero downtime if they invest work in doing a zero downtime transition. 
Downtime is "up to them"

Author: Let's not mention that IMO

</BranchSection>

<BranchSection path="wait">
**Wait for automatic**: You're done after Step 1. On July 31, the Pyth DAO upgrades your contract and `hermes.pyth.network` starts requiring authentication. You'll need to push a code change adding the `Authorization: Bearer $PYTH_API_KEY` header at that point. For a safer rollout, start using the authenticated Hermes API (`https://pyth.dourolabs.app/hermes`) before cutover as a fallback path. Plan a deploy around the cutover.
</BranchSection>

<BranchSection path="now">

## If you're upgrading early

<Steps>
<Step>

### Move your Hermes calls to the upgraded endpoint

> Switch your Hermes base URL from `hermes.pyth.network` to `pyth.dourolabs.app/hermes` and add your API key as a bearer token. The two URLs serve the same data today. The upgraded endpoint requires authentication.

Reviewer: I think mentioning Bearer tokens is too low level. Just say add your API key. If they want the low level detail they will get it from the curl example.

Reviewer: Suggest using API key instead of Bearer token throughout the doc.

Author: Fair.

<Tabs items={["curl", "TypeScript SDK"]}>
<Tab value="curl">

```diff
- url: "https://hermes.pyth.network"
+ url: "https://pyth.dourolabs.app/hermes"

> + Authorization: "Bearer $PYTH_API_KEY"

Reviewer: Let's include a full curl command here

```

</Tab>
<Tab value="TypeScript SDK">

```ts
import { HermesClient } from "@pythnetwork/hermes-client";

const client = new HermesClient("https://pyth.dourolabs.app/hermes", {
  headers: {
    Authorization: `Bearer ${process.env.PYTH_API_KEY}`,

>   },

Reviewer: Let's use `{accessToken: process.env.PYTH_API_KEY}` instead of manually setting the headers in the code examples

});
```

</Tab>
</Tabs>

<Callout type="info">

>   **Most resilient: use both URLs as primary + fallback.** Keeps you live

Reviewer: This logic only works post-cutover. It assumes both endpoints return the same data.

Reviewer: The correct logic would be something like:
- Get the data both from https://pyth.dourolabs.app/hermes and https://hermes.pyth.network.
-  Simulate the contract call with both payloads. 
- Keep the payload that succeeded simulation. 
- Make the contract call with that payload.

Author: Well, that's confusing then. Then we should ask them to keep one. Nobody will going to try with both.

</Callout>

Routes and response shapes are unchanged.
The upgraded endpoint is a drop-in replacement. See the [Hermes API reference](https://pyth.dourolabs.app/docs/?urls.primaryName=Hermes+API#/) for the full surface.

</Step>
<Step>

### Swap your contract address

Swap your existing Pyth contract address for the [upgraded Pyth Core Contract](/price-feeds/core/upgrade/contracts) on your chain. The upgraded Pyth Core Contract preserves the Pyth Core interface. No other code changes are needed.

View all upgraded [Pyth Core Contract addresses](/price-feeds/core/upgrade/contracts).

</Step>
</Steps>

</BranchSection>

## Chain support

Pyth Core will be supported on the chains listed on the [upgraded Pyth Core Contract addresses page](/price-feeds/core/upgrade/contracts). If your chain isn't listed, [contact the team](#get-help) — additional support is available by custom arrangement.

### Feed support

Nearly all current Pyth Core feeds remain available after the upgrade, with new ones added. Look up your specific feeds on the [feed explorer](https://pythdata.app/explore). If a feed you depend on isn't listed, [contact the team](#get-help).

## FAQ

<Accordions type="single">

>   <Accordion title="What happens if I do nothing?">

Reviewer: If you do nothing your hermes API calls will start getting rejected at the cutover because you are not using hermes with an API key.

    On July 31, 2026, the Pyth DAO upgrades your contract for you and `hermes.pyth.network` starts requiring authentication. Two things you still have to do:

    1. Have a Pyth API Key (Step 1) by that date.

>     2. Push a code change adding `Authorization: Bearer $PYTH_API_KEY` to your Hermes requests, around the cutover. Until that deploy lands, Hermes calls will fail.

Reviewer: The Authorization HTTP header is too low level and is only relevant if they hand-roll their Hermes API calls.

If you want to speak high level use the word: "API key"
If you want to speak about using our published Hermes clients use the term "access token"

---
title: "Preparing for the Pyth Core upgrade"
description: "Everything you need to upgrade your Pyth Core integration before July 31, 2026."
icon: RocketLaunch
full: true
---

import { Button } from "@pythnetwork/component-library/Button";
import { Accordion, Accordions } from "fumadocs-ui/components/accordion";
import { Callout } from "fumadocs-ui/components/callout";
import { Card, Cards } from "fumadocs-ui/components/card";
import { Step, Steps } from "fumadocs-ui/components/steps";
import { Tab, Tabs } from "fumadocs-ui/components/tabs";
import {
  ActiveStepHighlighter,
  BranchSection,
  BranchToggle,
} from "../../../../../src/components/MigrationFlow";

<ActiveStepHighlighter />

## What's changing

Pyth Network is upgrading [Pyth Core](/price-feeds/core) on **July 31, 2026**.
This upgrade replaces Pyth Core's underlying data infrastructure with an improved version, enabling new features:

- Higher-frequency updates
- Customizable channels

> - Access to Pyth Terminal

Reviewer: I would replace this with some stronger points from our blog that are clearly ways that the product gets better:
- Higher-frequency updates
- Additional price feeds
- Lower latency

Terminal access and Customizable channels are not really improvements in my opinion

  variant="primary"
>
  Sign up at Pyth Terminal
</Button>

### Step 2: Upgrade now or wait for automatic?

After Step 1, choose how and when to move to the upgraded infrastructure. Your choice is saved in the URL so you can share or bookmark a specific path.

<BranchToggle />

|                  | Upgrade now (recommended)                  | Wait for automatic                                                   |
| ---------------- | ------------------------------------------ | -------------------------------------------------------------------- |
| Timing           | You choose                                 | July 31, 2026                                                        |
| Downtime         | Up to you                                       | Brief, during the switch                                             |
| Hermes endpoint  | Switch to the new Hermes endpoint now             | Start using `hermes.pyth.network` with an API key before the cutover               |
| Contract address | You swap to the upgraded Pyth Core Contract       | DAO upgrades the current Pyth Core Contract for you                             |

<BranchSection path="now">

## Upgrading Now

> This is the recommended path, as it makes the **July 31 cutover a non-event for your integration**. You can switch to the new endpoint and swap addresses in the same deploy.

Reviewer: I think you make it sound like it's so easy. In reality program upgrades are done via multisig and are hard to synchronize with offchain services deployments.

Reviewer: In fact, if they want to do zero-downtime, they'll probably need to implement the code below that tests payloads from boths URLs and picks the one that verifies.

The difference between Upgrade Now and Wait for automatic in my opinion is they choose when their contracts cutover happens and don't need to coordinate with us.

If they wait for automatic an external team (us) chooses the cutover and we may move the date around or get delayed because of the Pythian Council, you have more unpredictability.

> This is the recommended path, as it makes the **July 31 cutover a non-event for your integration**. You can switch to the new endpoint and swap addresses in the same deploy.

Reviewer: ```suggestion
This is the recommended path, as it makes the **July 31 cutover a non-event for your integration**. Please note that if you swap the endpoint but not the contract address or vice versa you won't be able to verify price updates. Make sure both changes are deployed simultaneously or implement a forward compatible strategy like the one described in the Automatic Upgrade section.
```

Reviewer: Feel free to split this between this paragraph and the warning at the end of [Move your Hermes calls to the new Hermes endpoint](https://developer-hub-git-docs-migration-hub-pyth-network.vercel.app/price-feeds/core/upgrade#move-your-hermes-calls-to-the-new-hermes-endpoint)

<Steps>
<Step>

### Move your Hermes calls to the new Hermes endpoint

> Switch your Hermes base URL from [`hermes.pyth.network`](https://hermes.pyth.network/docs/) to [`pyth.dourolabs.app/hermes`](https://pyth.dourolabs.app/docs/?urls.primaryName=Hermes+API) and add your API key that you got in Step 1. The two URLs serve the same data today. The upgraded endpoint requires authentication.

Reviewer: Wrong. The two endpoints don't serve the same data.

Author: What's the difference. As far as I know the new endpoint have different fields, but the id is same. I think that matters the most.

Reviewer: I still feel like you should fix this: see https://github.com/pyth-network/pyth-crosschain/pull/3742#discussion_r3280614668

<Tabs items={["curl", "TypeScript SDK"]}>
<Tab value="curl">

```bash copy
curl -H "Authorization: Bearer $PYTH_API_KEY" \
  "https://pyth.dourolabs.app/hermes/v2/updates/price/latest?ids[]=0xe62df6c8b4a85fe1a67db44dc12de5db330f7ac66b72dc658afedf0f4a415b43"
```

</Tab>
<Tab value="TypeScript SDK">

```ts
import { HermesClient } from "@pythnetwork/hermes-client";

const client = new HermesClient("https://pyth.dourolabs.app/hermes", {
  accessToken: process.env.PYTH_API_KEY,
});
```

</Tab>
</Tabs>

Routes and response shapes are unchanged.

> The upgraded endpoint is a drop-in replacement. See the new [Hermes API reference](https://pyth.dourolabs.app/docs/?urls.primaryName=Hermes+API#/) for the full surface. Note that this is the new endpoint, not the old one.

Reviewer: ```suggestion
The upgraded endpoint is a drop-in replacement. See the new [Hermes API reference](https://pyth.dourolabs.app/docs/?urls.primaryName=Hermes+API#/) for the full surface.
```
That sentence didn't have that much of a purpose imo.

<Callout type="warn">
  **Pair this with Step 2 in the same deploy.** The new endpoint serves router-signed payloads that only the upgraded Pyth Core Contract verifies.
</Callout>

</Step>
<Step>

### Swap your contract address

Swap your existing Pyth contract address for the [upgraded Pyth Core Contract](/price-feeds/core/upgrade/contracts) on your chain. The upgraded Pyth Core Contract preserves the Pyth Core interface. No other code changes are needed.

View all upgraded [Pyth Core Contract addresses](/price-feeds/core/upgrade/contracts).

After the swap, you're done before the July 31 cutover. 

</Step>
</Steps>

</BranchSection>

<BranchSection path="wait">

## Waiting for the Automatic Upgrade

At the cutover on July 31, the Pyth DAO upgrades your on-chain contract a few moments *before* `hermes.pyth.network` starts serving the upgraded payload.

> During that brief gap, `hermes.pyth.network` cannot verify against your upgraded contract, while `pyth.dourolabs.app/hermes` can.

Reviewer: ```suggestion
During that brief gap, the price updates served by `hermes.pyth.network` cannot verify against your upgraded contract, while those served by `pyth.dourolabs.app/hermes` can.
```

> To stay live without timing the switch yourself, fetch from both URLs every request and submit whichever payload verifies:

Reviewer: ```suggestion
A potential way to stay live without timing the switch yourself is fetching from both URLs every request and submit whichever payload verifies:
```
maybe they can come up with some other ideas that apply to their setups, we don't need to be too prescriptive here. 
It's more about giving some ideas and design space

  try {
    await pythContract.simulate.updatePriceFeeds([update.binary.data], { value: fee });
    await pythContract.write.updatePriceFeeds([update.binary.data], { value: fee });
    break;
  } catch {
    // doesn't verify yet — try the other one
  }
}
```

</BranchSection>

## Chain support

Pyth Core will be supported on the chains listed on the [upgraded Pyth Core Contract addresses page](/price-feeds/core/upgrade/contracts). If your chain isn't listed, [contact the team](#get-help). Additional support is available by custom arrangement.

### Feed support

Nearly all current Pyth Core feeds remain available after the upgrade, with new ones added. Look up your specific feeds on the [feed explorer](https://pythdata.app/explore). If a feed you depend on isn't listed, [contact the team](#get-help).

## FAQ

<Accordions type="single">
  <Accordion title="What happens if I do nothing?">
    **Your Hermes calls will start getting rejected at the cutover on July 31, 2026** — `hermes.pyth.network` will require an API key from that point on. To stay live, please follow the steps mentioned [above](#waiting-for-the-automatic-upgrade).
  </Accordion>

>   <Accordion title="Will my existing code break?">

Reviewer: Maybe a better answer here.

If you sign up for an API key and start using Hermes with it, nothing will break. You may however experience a brief oracle downtime on the cutover if you don't take proactive steps to avoid it.

  </Accordion>
  <Accordion title="Are all my feeds available after the upgrade?">
    Nearly all current Pyth Core feeds remain available after the upgrade, with new ones added. Look up your specific feeds on the [feed explorer](https://pythdata.app/explore). If a feed you depend on isn't listed, [contact the team](#get-help).
  </Accordion>
  <Accordion title="I read prices on-chain only (push feeds, no Hermes) — do I need an API key?">
    No. The API key is only required for Hermes (REST/SSE) requests. If your integration reads prices directly from the Pyth contract on-chain and never calls Hermes, you don't need to register. Your contract will be upgraded automatically on July 31, or you can swap it early (Step 3).
  </Accordion>
  <Accordion title="My code uses the Hermes REST/SSE API directly — what changes for me?">
    Routes and response shapes are unchanged. The upgraded endpoint is a drop-in replacement. The differences are *which URL* you call and *whether you send an API key*:

>     - **Upgrade early (recommended):** switch your base URL to `pyth.dourolabs.app/hermes` and add `Authorization: Bearer $PYTH_API_KEY` header today. The data is identical; only auth changes. See the [upgrade guide](/price-feeds/core/upgrade) for the full path.

Reviewer: ```suggestion
    - **Upgrade early (recommended):** switch your base URL to `pyth.dourolabs.app/hermes` and start attaching and API key to your requests. The data is identical; only auth changes. See the [upgrade guide](/price-feeds/core/upgrade) for the full path.
```

>     - **Wait:** stay on `hermes.pyth.network` today (unauthenticated). On July 31, push a deploy that adds the API key header on the same URL.

Reviewer: wrong! You need to add the API asap. Don't wait for the last day to get the API key and starting adding the header.
